### ISSUE TITLE
Support for quantization/quantized models?

### ISSUE THREAD

**Issue by pbatk** [guillaume-be/rust-bert#124]
Hi,

Great work on this library! It is awesome to work with!

I was hoping to get some insight on whether you had considered quantization for this library. Seeing as how rust would be used mostly for production grade systems and quantization is a production-time solution it seems like a logical next step.

Thoughts?

**Comment by guillaume-be** [comment]
Hello @pbatk ,

Thank you for your interest in the library. Indeed quantization is an interesting and natural next step. I did look into the implementation of state-of-the-art quantized models such as [TernaryBERT](https://arxiv.org/abs/2009.12812) but I haven't had the opportunity to experiment with it yet.

In general there are multiple mechanism to run inference with quantization:
- Dynamic quantization seems to be one of the most flexible for the Pytorch library. I am not aware of an implementation of the [quantize_dynamic](https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.quantize_dynamic) for the Libtorch C++ API and therefore the [Rust bindings](https://github.com/LaurentMazare/tch-rs). I don't believe this would be the most promising implementation path.
- Static quantization (either with quantized-aware training or not) stores the weights and uses activations in a quantized format. I believe this may work with some modifications (checking the type of the tensors dynamically) of the models. I haven't had the opportunity to test it: there are no models saved in a quantized format in the Transformers' model hub. 
- Torchscript JIT is probably another promising alternative. The Rust bindings do support the loading of JIT quantized models (see https://github.com/LaurentMazare/tch-rs/tree/master/examples/jit-quantized). One challenge with tracing transformers models is linked to the complex layers that may involve conditional statements. The reference Transformers library still marks this feature as experimental: https://huggingface.co/transformers/torchscript.html. Please note also that the sequence length becomes constant after export to a torchscript. This means that while you can expect speed benefits from the quantization, shorter inputs will be unnecessarily padded to the sequence length used for tracing which will in turn result in increased computation cost. Please note also that loading these torchscript models does not require the implementation proposed in this crate (as illustrated in https://github.com/LaurentMazare/tch-rs/tree/master/examples/jit-quantized)

Do you already have a model in mind for quantization? We could then use it as a basis for these investigations

**Comment by pbatk** [comment]
Thanks for the time and thought!
My immediate needs would be the Marian models. However, since I have begun to work with transformers in Rust instead of Python I have become largely convinced that this will be my production language for transformers. 

I would be happy to help implement the quantization in my spare time. With that said, I really only have about 4-6 hours a week so this will likely need to be a slow and steady implementation

**Comment by pbatk** [comment]
The quantize_dynamic and Torchscript solutions interest me most. Personally the dynamic option is the implementation I envision however I do not know if it would be reasonable to implement this from scratch in native Rust (I will investigate the Python/C++ implementations).

Torchscript seems like a reasonable middle ground but the static length is concerning. Most of my transformer use cases take advantage of dynamic length for extra performance. From your perspective, how time consuming do you believe a Torchscript implementation using tch-rs would be (disregarding the performance and experimental nature of it)?

**Comment by guillaume-be** [comment]
Hello,

I would be happy to help wherever I can - please note that I also have limited time to devote to this project (4-6 hours weekly as well). 

Dynamic quantization indeed seems to be one of the most flexible schemes for implementing quantization. However, looking at the Python implementation I believe significant efforts would be required: the conversion loops through the modules of a target Pytorch model and replaces the standard modules (e.g. all `nn.Linear`) by their quantized version (e.g. `nn.quantized.Linear`). While this works well for a dynamically typed language like Python, I see significant challenges with the Rust implementation and static typing of the modules used.

The static quantization may be an option with less required changes, but would probably require calibration and quantization aware training of the models to avoid performance degradation.

The Torchscript implementation may be easier to implement in general. However, text generation models, such as BART and Marian, are likely going to be more tricky because of their complex input format and the use of cache. There may be easier targets (for example a sequence classification task) to experiment with quantization without the added complexity of caching and the text generation routine.

In general - regardless of the way forward, this is probably a fairly complex feature to implement - and I am not sure I will have the capacity to work on this right away. I'd be happy for you to look in this implementation and support wherever I can. 

I would have a few hints to get started:
- Do you have a target model in mind? It may be worth having a reference working workflow in Python to identify the most promising quantization scheme for your application (Do you already have a quantized Marian pipeline working in Python?)
- I'd reach out to the `tch-rs` maintainers to gather further information. Examples are provided with torchscript (although for image where tracing tends to be easier than for language models), but it would be good to know if dynamic or static quantization without Torchscript is feasible.

**Comment by pbatk** [comment]
@guillaume-be Just to update you on my plans. I have not yet dedicated time to this but plan to begin next week. I will likely take a week to formulate my thoughts and a plan. I will update then.

For now I am going to close this as it is more a long term feature request than a current issue. I will continue to use this as a communication forum though

**Comment by AlexMikhalev** [comment]
It may be worth thinking of integrating quantized methods from ONNX https://github.com/microsoft/onnxruntime .

**Comment by pbatk** [comment]
@AlexMikhalev I'll check it out. Unfortunately other priorities have taken over at work and left me without the extra time to work on this. It is still important to us so I'll eventually loop back around
